source/game/kart/KartDynamics.cc

```diff
@@ -7,34 +7,59 @@ KartDynamics::KartDynamics() {
 }
 
 void KartDynamics::init() {
-    m_pos.setZero();
-    m_speed.setZero();
-    m_fullRot = EGG::Quatf(1.0f, EGG::Vector3f::zero);
+    m_speedNorm = 0.0f;
+    m_gravity = -1.0f;
     m_top = EGG::Vector3f::ey;
+    m_forceUpright = true;
+    m_noGravity = false;
 }
 
-const EGG::Vector3f &KartDynamics::getPos() const {
+void KartDynamics::resetInternalVelocity() {
+    m_velocity.setZero();
+}
+
+void KartDynamics::calc(f32 dt, f32 /*maxSpeed*/, bool /*air*/) {
+    if (!m_noGravity) {
+        m_totalForce.y += m_gravity;
+    }
+
+    m_acceleration0 = m_totalForce;
+    m_externalVel += m_acceleration0 * dt;
+    m_externalVel *= 0.998f;
+    m_velocity =
+            m_externalVel * dt + m_internalVelocity + m_movingObjectVelocity + m_movingRoadVelocity;
+    m_speedNorm = m_velocity.normalise();
+    m_velocity *= m_speedNorm;
+    m_pos += m_velocity;
+
+    m_fullRot = m_extraRot * m_mainRot * m_specialRot;
```

Comment:
We can prove this triple-quaternion multiplication happens in the intended order, right?

Reply:
It occurs in the intended order.
![aa3b82df5533d9d485e20e060f1ac101](https://github.com/vabold/Kinoko/assets/16770560/b5e57b3b-37ee-48ea-bc62-b07731aa5c2a)